### PR TITLE
Dockerize WSIMOD

### LINKED ISSUE #25: Create docker container for WSIMOD
After #24 , it should be relatively easy to create a Dockerfile that let you run WSIMOD within a docker container. A `docker-compose.yml` file should also be added to simplify its local usage and testing. These files could live within a `dafni` subfolder in the main repository - or directly in the root folder if they are just those two.

The `Dockerfile` should be written such that it creates an image that, when executed, will run `wsimod` CLI with the relevant input arguments:

```
python -m wsimod --settings /data/inputs/settings.toml`
```

It will trust that, when run, the correct volumes are mounted in the container (namely that `/data/inputs` and `/data/outputs` are available) and that the data is available where it is supposed to be, according to the `settings.toml` file.

While we will do our best to make this file as general as possible, its goal its to confirm to the requirements of the DAFNI framework, so in case of doubt, it will be tailored specifically to fulfill its needs. https://docs.secure.dafni.rl.ac.uk/docs/how-to/how-to-create-a-dafni-ready-model

### PR DESCRIPTION
**Pull request by dalonsoa**
This PR dockerize the execution of WSIMOD and adds a `docker-compose` file to facilitate local testing and deployment. 

To be reviewed and merged after #38 , as it builds on top of it. 

Closes #25

### CODE REVIEW

**Comment by dalonsoa** [comment]
To test the code:

#### Without Docker
Just install WSIMOD and run `wsimod docs/demo/examples/quickstart_demo.yaml`. There should be a progress bar and a few files generated in `results/quickstart_results`.

#### With Docker
This is meant to run in DAFNI, so the file structure needs to be a little specific. 

1. Create a folder to host your **input** data wherever you want.
2. Create a folder to host you **output** data wherever you want.
3. Copy `docs/demo/data/processed/timeseries_data.csv` to the inputs folder.
4. Copy `docs/demo/examples/quickstart_demo.yaml` to the inputs folder.
5. Create a `.env` file in the root directory of the project for `docker-compose` to read and put the following environment variables inside:
```text
WSIMOD_SETTINGS={name_of_yaml_file_with_extension}
WSIMOD_INPUTS={root_directory_for_the_inputs}
WSIMOD_OUTPUTS={root_directory_for_the_outputs}
```
For reference, this is what I have, with the inputs and outputs in the same folder, the the yaml file called `settings_new.yaml`:
```text
WSIMOD_SETTINGS=settings_new.yaml
WSIMOD_INPUTS=./scratch
WSIMOD_OUTPUTS=./scratch
```
6. Run `docker-compose up`

The output should be conceptually similar to the previous case, getting the same output files in the output folder. 

@AdrianDAlessandro , @TinyMarsh , @tsmbland , @tg2414 , in case you want to try this.

**Review by TinyMarsh — APPROVED** [review]
LGTM

**Review by tsmbland — APPROVED** [review]

File: docker-compose.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+version: '3'
+services:
+  wsimod:
+    build: .
+    env_file:
+      - .env
```
Comment (AdrianDAlessandro) [inline docker-compose.yaml:6]: I believe this is redundant because docker compose uses `.env` by default

**Review by AdrianDAlessandro — COMMENTED** [review]
I don't understand why you need to move things around and create files in different locations. Why not just change where those files are in the repo so that it is compatible with DAFNI? Removing as much manual intervention as possible is usually the best way to go here.

Also, in the instructions it's not clear which file `WSIMOD_SETTINGS` is meant to be. You use a file called `settings_new.yaml` but I don't know where to get that from / make it.

Also, the docker build failed for me, it failed to build wheels for `numpy==1.26.2` - this might be because that version of numpy is too recent for the docker image, I'm not sure. It successfully installed in a virtual environment.

**Comment by dalonsoa** [comment]
Keeping aside the WSIMOD code, there're no files in the repo needed for DAFNI. The files I copy are just taken from the documentation/tutorials, using them as an example of how to run the tool. Neither the `Dockerfile` nor the `docker-compose.yaml` file need those files specifically, just some input files to run the model with as an example. Therefore, calling a settings file `settings.yaml`, `settings_new.yaml` or `something_different.yaml` is irrelevant as long as that name is passed to the container. In the instructions, that file is `quickstart_demo.yaml`. I've updated the expected contents of the `.env` file to reflect this.

Those input files (the settings as well as any required data file) will need to be provided by the user in DAFNI using the DAFNI web interface, so nothing for us to worry about.

About the error with `numpy`, that's very odd because I'm using exactly that version - pinned in the `requirements.txt` file -, with that particular base image, and it works without any problem. No idea where your problem might be coming from. What architecture does a Docker container see? It might be that for Python:3.11.5-alpine numpy does not exist for ARM processors?

On your last comment, you need to include explicitly the environment file to use if you need the environment variables in the `docker-image.yaml` file itself, which is the case. Without that section, `docker-compose` fails saying it cannot find `"WSIMOD_INPUTS"` and `"WSIMOD_OUTPUTS"`

File: docker-compose.yaml (hunk 1 above)
Comment (dalonsoa) [reply docker-compose.yaml:6]: Apparently it is not used automatically, no.

Comment (AdrianDAlessandro) [reply docker-compose.yaml:6]: Oh yes true, it uses it automatically for setting environment variables within the container, but not the way you're using them here

**Review by AdrianDAlessandro — APPROVED** [review]
Ok great, that all makes sense. As long as you can build the docker image, it's fine. I think my issue is probably ARM-related

**Pull request merged**